packages/integration-react/src/api/ScmAuth.ts

```diff
@@ -0,0 +1,235 @@
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OAuthApi } from '@backstage/core-plugin-api';
+import {
+  ScmAuthApi,
+  ScmAuthTokenOptions,
+  ScmAuthTokenResponse,
+} from './ScmAuthApi';
+
+type ScopeMapping = {
+  default: string[];
+  repoWrite: string[];
+};
+
+class ScmAuthMux implements ScmAuthApi {
+  #providers = new Array<ScmAuth>();
+
+  constructor(providers: ScmAuth[]) {
+    this.#providers = providers;
+  }
+
+  async getCredentials(
+    options: ScmAuthTokenOptions,
+  ): Promise<ScmAuthTokenResponse> {
+    const url = new URL(options.url);
+    const provider = this.#providers.find(p => p.isUrlSupported(url));
+    if (!provider) {
+      throw new Error(
+        `No authentication provider available for access to '${options.url}'`,
+      );
+    }
+
+    return provider.getCredentials(options);
+  }
+}
+
+/**
+ * An implementation of the ScmAuthApi that merges together OAuthApi instances
+ * to form a single instance that can handles authentication for multiple providers.
+ */
+export class ScmAuth implements ScmAuthApi {
+  /**
+   * Creates a general purpose ScmAuth instance with a custom scope mapping.
+   */
+  static forAuthApi(
+    authApi: OAuthApi,
+    options: {
+      host: string;
+      scopeMapping: {
+        default: string[];
+        repoWrite: string[];
+      };
+    },
+  ): ScmAuth {
+    return new ScmAuth(authApi, options.host, options.scopeMapping);
+  }
+
+  /**
+   * Creates a new ScmAuth instance that handles authentication towards GitHub.
+   *
+   * The host option determines which URLs that are handled by this instance and defaults to `github.com`.
+   *
+   * The default scopes are:
+   *
+   * `repo read:org read:user`
+   *
+   * If the additional `repoWrite` permission is requested, these scopes are added:
+   *
+   * `gist`
+   */
+  static forGithub(
+    githubAuthApi: OAuthApi,
+    options?: {
+      host?: string;
+    },
+  ): ScmAuth {
+    const host = options?.host ?? 'github.com';
+    return new ScmAuth(githubAuthApi, host, {
+      default: ['repo', 'read:org', 'read:user'],
+      repoWrite: ['repo', 'read:org', 'read:user', 'gist'],
+    });
+  }
+
+  /**
+   * Creates a new ScmAuth instance that handles authentication towards GitLab.
+   *
+   * The host option determines which URLs that are handled by this instance and defaults to `gitlab.com`.
+   *
+   * The default scopes are:
+   *
+   * `read_user read_api read_repository`
+   *
+   * If the additional `repoWrite` permission is requested, these scopes are added:
+   *
+   * `write_repository api`
+   */
+  static forGitlab(
+    gitlabAuthApi: OAuthApi,
+    options?: {
```

Comment:
I was just curious: What is the reasoning behind just having `default` and `repoWrite` scopes? Or is this just intended as an initial mapping?

For instance, the GitLab API has a `sudo` scope which, if the user has it, can be used to impersonate other users for a specific API call.

It seems like that could work in this new API approach by passing in the sudo username in the `options` object and then, if it's set, also automatically add the `sudo` scope to the requested scopes below. But then, how to determine as a Backstage developer which options (like `host` and `sudo`) should be set during the `forGitlab` call, and which options (like `repoWrite`) are added during the `getCredentials` call?

Reply:
The idea is to provide a unified API that provides a similar level of access to all auth providers. The reason for splitting it into read/write was for security, but I didn't wanna get more granular than that because the way scopes work vary pretty wildly between the providers.

Individual options for things like GitLab's `sudo` kind of defeats the purpose a bit. It might be that we need to add more options in the future, but initially I want to keep it very slim.

Reply:
- @Rugvip do you think it makes sense to also include `ownerWrite` as well to handle owner/org/project/workspace auth? 

We have a need to extend the RepoUrlPicker to have a dropdown where only projects the user has access is shown, once a project or org is selected they would either select an existing repository or enter the name of the new repository they would like to create

Reply:
The default scopes should give read information to that data. I don't think we need to write there at this point right?

Reply:
If that's the case we should be good!